### PR TITLE
EQL Samples: add support for multiple samples per key

### PR DESCRIPTION
**Pull request by luigidellaquila**
By default EQL Samples queries return only one sample per key.
With this PR we add a `max_samples_per_key` option, that allows to return multiple samples for the same key.

eg. 

```
  sample by os with max_samples_per_key=2
    [alert where true]
    [alert where true]
```

[edit] the final implementation uses a request body parameter to pass max_samples_per_key:

```
{
   "max_samples_per_key": 10,
   "query": "sample by os [..."
}
```

### CODE REVIEW

File: x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sample/SampleIterator.java [hunk 1]
```diff
@@ -84,13 +85,21 @@ public class SampleIterator implements Executable {
...
+        CircuitBreaker circuitBreaker,
+        int maxSamplesPerKey
+    ) {
         this.client = client;
         this.criteria = criteria;
         this.maxCriteria = criteria.size();
         this.fetchSize = fetchSize;
         this.samples = new ArrayList<>();
         this.limit = limit;
+        this.maxSamplesPerKey = maxSamplesPerKey;
```
Comment (astefan) [inline x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sample/SampleIterator.java:102]: After `circuitBreaker`, please. Keep the same order as the one of the constructor's arguments.

**Review by astefan — CHANGES_REQUESTED** [review]
This type of characteristic/setting/query hint I don't think it fits the `with` construct in the language. On one hand, `with` is used in combination with `maxspan` indicating the period of time the events in a sequence should fit into.
`with` is also with `runs` being more like a syntactic sugar for repeating the same condition more times.

On the other hand, EQL has body parameters like `fetch_size` (a setting that's being used internally by the sequence algorithm), `result_position` (changes the default recency aspect of events - earlier or later events to return) and `size` (the maximum number of events or sequences to return).

From the two available options above (inside the language itself or inside the body of the request), I think `max_samples_per_key` fits better in the body of the request because it's an aspect of the query that is not influencing the way the query runs, but more how the results are returned.

File: x-pack/plugin/eql/qa/common/src/main/resources/data/sample-multi.data [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    "host": "GTA",
+    "os": "fedora",
+    "bool": true,
+    "id": 5
+  },
+  {
+    "event_type": "alert_2",
+    "host": "farcry",
+    "bool": true,
+    "os": "reahat",
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/eql/qa/common/src/main/resources/data/sample-multi.data:42]: Just a typo (copy-pasted), fixing it with next commit.

**Comment by astefan** [comment]
> @astefan thank you very much for your feedback. I thought a bit about it, but I dont' think I completely agree with your conclusions. I'll leave my observations here for further discussion:
> 
> * `with` is the way we define extensions to the language, so it seems the natural way to proceed.

It's not an extension where we can simply pop things in whenever the language needs a feature or setting. The construct using `with` needs to have a meaning, to be consistent with the way the language "feels" and works. And, as I mentioned in my review, there are aspects that break the consistency of the language.

> * The request body is only for transport, it should not contain query logic (the query logic should be in the EQL query itself).
>   Parameters like `fetch_size` are internal and can be considered as tuning options, while `max_samples_per_key` defines a query criterion (should one sample be returned or not?), so it's strict business logic.

I don't find this argument strong enough tbh.
> `max_samples_per_key` defines a query criterion (should one sample be returned or not?)

A query criterion (at language level) refers to something else and this setting is not defining that.

> so it's strict business logic

So is `size` and it's a request body parameter.

> * The REST API is one possible protocol, but not necessarily the only one (think of Endgame EQL CLI, today it doesn't support these features, but in future it could).
>   The language should support the feature regardless of the protocol.

There are features and characteristics that are different in these languages and each of them is making the best to accommodate the missing functionality but the use cases and the technical differences between them (Endgame EQL deployed as an agent that acts on a smaller data set realtime, ES EQL using ES to search a far larger dataset) make this goal to be complex and sometimes impossible to fulfill.

> * Moreover, the request body is generic for all the EQL queries, while `max_samples_per_key` is specific for Samples.
>   If we add it to the body, we will have to ignore it or to explicitly check it for all the other queries (eg. sequences) and throw errors.

`fetch_size` is a similar setting only used for sequences.

> * From a usability point of view, I think it's easier to define the query logic in EQL directly. The JSON body is much closer to machine protocol, it should be manipulated by the user as little as possible.
> * In some contexts, the end user could just not have access to the request body. Eg. Kibana dashboards provide a simple text area to write plain EQL. In this context, the user won't be able to use this feature.

I agree this is more user-friendly to be part of the language, but at the same time the construct (as I mentioned previously) needs to make sense and to be consistent with the rest of the language.

**Comment by sodhikirti07** [comment]
@luigidellaquila In my opinion, keeping the `max_samples_per_key` under the body of the `sample` query makes more sense. 

> In some contexts, the end user could just not have access to the request body. Eg. Kibana dashboards provide a simple text area to write plain EQL. In this context, the user won't be able to use this feature.

To my knowledge, I've only seen EQL being used for detection-rules in Kibana. In that case, I don't think we would need to set this parameter as the alerts are governed by time intervals. So any alert under an interval of 5m (let's say) would  show up under the alerts page.

**Comment by luigidellaquila** [comment]
@sodhikirti07 do you mean the body of the JSON, eg.  

```
{ 
   "max_samples_per_key":10, 
   "query": "sample [ ... ]"}
```

or in the query language, eg.

```
{ 
   "query": "sample with max_samples_per_key=10 [...]"}
```


Sorry, we had problems with names and definitions as well...

**Comment by sodhikirti07** [comment]
I would say the first one similar to what we do in case of `size`. 

```
{ 
   "max_samples_per_key":10, 
   "query": "sample [ ... ]"}
```

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by luigidellaquila** [comment]
@elasticmachine test this please

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sample/SampleIteratorTests.java [hunk 3]
```diff
@@ -15,24 +15,38 @@
...
+
+        // multiple matches per key
+        assertEquals(
+            List.of(asSearchHitsList(1, 3, 4), asSearchHitsList(1, 4, 3)),
+            matchSamples(asList(asSearchHitsList(1, 2, 3), asSearchHitsList(1, 3, 4), asSearchHitsList(1, 3, 4)), 3, 2)
+        );
+
+        assertEquals(
+            List.of(asSearchHitsList(1, 3, 4), asSearchHitsList(1, 4, 3), asSearchHitsList(2, 1, 3)),
+            matchSamples(asList(asSearchHitsList(1, 2, 3), asSearchHitsList(1, 3, 4), asSearchHitsList(1, 3, 4)), 3, 3)
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sample/SampleIteratorTests.java:48]: These two tests are not enough. Incrementally increase maxSamplesPerKey to the maximum possible value and test the combinations please.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/OptimizerTests.java [hunk 4]
```diff
@@ -794,7 +794,7 @@ private static Sequence sequence(LogicalPlan... rules) {
         return new Sequence(EMPTY, collect, keyedFilter(rel()), TimeValue.MINUS_ONE, timestamp(), tiebreaker(), OrderDirection.ASC);
     }
 
-    private static Sample sample(LogicalPlan... rules) {
+    private static Sample sample(int maxSamplesPerKey, LogicalPlan... rules) {
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/OptimizerTests.java:797]: Leftover.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/OptimizerTests.java [hunk 5]
```diff
@@ -723,7 +723,7 @@ public void testSampleOptimizations() {
 
     private AbstractJoin randomSequenceOrSample(KeyedFilter rule1, KeyedFilter rule2) {
         Sequence seq = sequence(rule1, rule2);
-        Sample sample = sample(rule1, rule2);
+        Sample sample = sample(randomIntBetween(1, 10), rule1, rule2);
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/OptimizerTests.java:726]: Leftover

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/OptimizerTests.java [hunk 6]
```diff
@@ -703,7 +703,7 @@ public void testReduceBinaryComparisons() {
 
     public void testSampleOptimizations() {
         String q = "sample by user_name [any where true] by bool [any where true] by bool";
-        LogicalPlan plan = sample(accept(q));
+        LogicalPlan plan = sample(randomIntBetween(1, 10), accept(q));
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/OptimizerTests.java:706]: Leftover

File: x-pack/plugin/eql/src/test/resources/queries-supported.eql [hunk 7]
```diff
@@ -732,3 +732,7 @@ sample by ?x, ?y, z, ?t [any where true] by z [any where t == y] by t
 sample by z, ?t [any where true] by ?a, ?b [any where t == y] by ?a, ?b
 sample by z, ?t [any where true] by ?a, ?b [any where t == y] by ?a, ?b
 sample by z, ?`t?t` [any where ?a == concat(?b,?c)] by ?a, ?b [any where ?`abc` == y] by ?a, ?b
+
+sample with max_samples_per_key=3 [any where true] [any where true]
+sample by x with max_samples_per_key=5 [any where true] [any where true]
+sample by ?x with max_samples_per_key=6 [any where true] by y [any where true] by z
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/resources/queries-supported.eql:738]: Leftover

**Review by astefan — COMMENTED** [review]
Left a first series of comments.

File: x-pack/plugin/eql/qa/common/src/main/resources/data/sample-multi.data [hunk 8]
```diff
@@ -0,0 +1,52 @@
+[
```
Comment (astefan) [inline x-pack/plugin/eql/qa/common/src/main/resources/data/sample-multi.data:1]: Add some empty join key values (`null`) or missing values and add some test queries that use optional fields as join keys. `sample by ?os [any where true] [any where true]`

File: x-pack/plugin/eql/qa/common/src/main/resources/test_sample_multi.toml [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+'''
+# non-deterministic, can return any of the following
+expected_event_ids  = [
+    [1, 3, 2, 3],
+    [1, 3, 4, 3],
+    [2, 3, 1, 3],
+    [2, 3, 4, 3],
+    [4, 3, 1, 3],
+    [4, 3, 2, 3]
+]
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/eql/qa/common/src/main/resources/test_sample_multi.toml:77]: Yes, it's all the possible combinations of N results per key.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by szeitlin** [comment]
yay! Thank you! Looking forward to trying this out!

**Pull request merged**